cli/internal/fs/copy_file.go

```diff
@@ -13,12 +13,16 @@ import (
 
 // CopyOrLinkFile either copies or hardlinks a file based on the link argument.
 // Falls back to a copy if link fails and fallback is true.
-func CopyOrLinkFile(from, to string, fromMode, toMode os.FileMode, link, fallback bool) error {
+func CopyOrLinkFile(from *LstatCachedFile, to string, link bool, fallback bool) error {
 	if link {
+		fromMode, err := from.GetMode()
+		if err != nil {
+			return err
+		}
 		if (fromMode & os.ModeSymlink) != 0 {
 			// Don't try to hard-link to a symlink, that doesn't work reliably across all platforms.
 			// Instead recreate an equivalent symlink in the new location.
-			dest, err := os.Readlink(from)
+			dest, err := os.Readlink(from.Path.ToString())
```

Comment:
Can you implement `fs.AbsolutePath.Readlink` and call that instead? You can see other models in `fs/path.go`. That will help us ensure that all of our filesystem interactions are via an absolute path.